idaes/core/util/model_diagnostics.py

```diff
@@ -4122,3 +4266,297 @@ def _collect_model_statistics(model):
+            yield abs(sum(i))
+
+    def _check_sum_cancellations(self, values_list):
+        # First, strip any terms with value 0 as they do not contribute to cancellation
+        # We do this to keep the number of possible combinations as small as possible
+        stripped = [i for i in values_list if i != 0]
+
+        if len(stripped) == 0:
+            # If the stripped list is empty, there are no non-zero terms
+            # We can stop here and return False as there are no possible cancellations
+            return False
+
+        # For scaling of tolerance, we want to compare to the largest absolute value of
+        # the input values
+        max_value = abs(max(stripped, key=abs))
+
+        for i in self._generate_sum_combinations(stripped):
+            # We do not need to check all combinations, as we only need determine if any
+            # combination cancels. Thus, if we find a cancellation, stop and return True
+            if i <= self._sum_tol * max_value:
+                return True
+
+        return False
+
+    def _perform_checks(self, node, child_data):
+        # Perform checks for problematic expressions
+        # First, need to check to see if any child data is a list
+        # This indicates a sum expression
+        const = True
+
+        for d in child_data:
+            # We will check for canceling terms here, rather than the sum itself, to handle special cases
+            # We want to look for cases where a sum term results in a value much smaller
+            # than the terms of the sum
+            if self._check_sum_cancellations(d[0]):
+                self.canceling_terms.add(node)
+
+            # Expression is not constant if any child is not constant
+            if not d[1]:
+                const = False
+
+        # Return any problematic terms found
+        return const
+
+    def _check_base_type(self, node):
+        if isinstance(node, VarData):
+            const = node.fixed
+        else:
+            const = True
+        return [value(node)], const, False
+
+    def _check_equality_expression(self, node, child_data):
+        # (In)equality expressions are a special case of sum expressions
+        # First, need to negate one side of the expression
+        mdata = []
+
+        vals = []
+        for j in child_data[0][0]:
+            vals.append(-j)
+        mdata.append((vals, child_data[0][1]))
+        mdata.append(child_data[1])
+
+        # Next, call the method to check the sum expression
+        vals, const, _ = self._check_sum_expression(node, mdata)
+
+        # Next, we need to check for canceling terms.
+        # In this case, we can safely ignore expressions of the form constant = sum()
+        # We can also ignore any constraint that is already flagged as mismatched
+        # We will also ignore any constraints where a == b and neither a nor b are sum expressions
+        if not child_data[0][2] and not child_data[1][2]:
```

Comment:
I am... mildly irked about all of the magic numbers here. Perhaps it is more well-known than me what is inside `child_data`, but as someone who doesn't know, what is in `[0][0]` and why does it need to be negated?

Reply:
Seeing as I had to remind myself what these meant, the point is well taken. I have added some comments that hopefully explain what is happening.